go-controller/pkg/allocator/pod/pod_annotation.go

```diff
@@ -261,6 +261,56 @@ func validateStaticIPRequest(netInfo util.NetInfo, network *nadapi.NetworkSelect
 	return nil
 }
 
+var (
+	ErrIPFamilyMismatch = errors.New("requested IPs family types must match network's IP family configuration")
+)
+
+func validateIPFamilyMatchesNetwork(netInfo util.NetInfo, requestedIPs []*net.IPNet) error {
+	if !util.IsPreconfiguredUDNAddressesEnabled() {
+		return nil
+	}
+	if !netInfo.IsPrimaryNetwork() {
+		return nil
+	}
+	if netInfo.TopologyType() != types.Layer2Topology {
+		return nil
+	}
+
+	if len(requestedIPs) == 0 {
+		return nil
+	}
+
+	var ipv4Count, ipv6Count int
+	for _, ipNet := range requestedIPs {
+		if utilnet.IsIPv6CIDR(ipNet) {
+			ipv6Count++
+		} else {
+			ipv4Count++
+		}
+	}
```

Comment:
Check first if there are more thatn 2 ip addresses, so we fail fast without looping.